core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -303,15 +311,25 @@ class LogSegment(val log: FileRecords,
     // We may need to reload the max timestamp after truncation.
     if (maxTimestampSoFar >= 0)
       loadLargestTimestamp(readToLogEnd = true)
+    _nextOffset = math.max(baseOffset, math.min(_nextOffset, offset))
     bytesTruncated
   }
 
   /**
-   * Calculate the offset that would be used for the next message to be append to this segment.
+   * The offset that would be used for the next message to be appended to this segment
+   */
+  def nextOffset(): Long = {
+    if (_nextOffset == INVALID_NEXT_OFFSET)
+      _nextOffset = nextOffsetFromLog()
```

Comment:
Thanks for the quick comment @junrao.

Yes it is possible to do that, i.e. we keep the previous expensive implementation of `nextOffset()` which scans the last portion of the log, and implement `deleteLogStartOffsetBreachedSegments()` in such a way that the deletion of a segment depends on the baseOffset of the next segment. I am just not very sure the advantage of that approach as compared to the approach in the current patch. In comparison, that approach keeps nextOffset() to be expensive, and requires more complex implementation of `deleteLogStartOffsetBreachedSegments()` because it no longer re-use `deleteOldSegments(...)` to delete partition. The benefit of the approach is that we don't need to maintain `nextOffset` as state in `LogSegment`.

I am OK with the suggested approach. Let me do that then.